### DISCUSSION TITLE
textual inversion and dreambooth

### DISCUSSION THREAD

**Discussion by adrianpuiu** [seruva19/kubin#10] [General]
hi buddy, saw your work and looks great... 
you might also want to look at : https://github.com/TheDenk/Kandinsky-2-textual-inversion 

the guy is working on drambooth and textual inversion scripts. would be great to have them integrated in your UI

**Comment by seruva19** [comment]
Hello, thank you.
I saw your post on r/StableDiffusion about gradio app you made, nice to see more people getting involved into Kandinsky.
Regarding textual inversion training: I saw that repo and actually planning to use it (I mentioned it in README). The problem is, currently I only have access to GPU with 16Gb VRAM. And according to information provided by the author of the method, it requires 24 Gb card, so I won't even be able to test if it works :)
That said, I still want to add it to GUI, so that users with high-end GPUs might try it.

**Reply by adrianpuiu** [reply]
i tried it today on collab with 40Gb Vram but give me a lot of trouble after generating the class images. i got 2 24 gb vram cards back in romania . next week im going back home , maybe we can meet over the weekend and try to debug and work on something. ? I have collab plus subscription so we can test it there..  what do you say ?

**Reply by seruva19** [reply]
Yes, it's a good idea, we can try. 
Did you ask the author of the textual inversion scripts about trouble you had with their script? I am not an expert, just learning, so there is a chance I won't be able to help.

**Reply by adrianpuiu** [reply]
I have not asked him yet, stil trying to find my way around it. i'm same as you, trying to catch up and learn :))

**Reply by adrianpuiu** [reply]
update : don't try the notebooks , they don't work. finally I've run the python train_textual_inversion.py and it works...  `/usr/local/lib/python3.9/dist-packages/huggingface_hub/file_download.py:645: FutureWarning: `cached_download` is the legacy way to download files from the HF hub, please consider upgrading to `hf_hub_download`
  warnings.warn(
making attention of type 'vanilla' with 512 in_channels
making attention of type 'vanilla' with 512 in_channels
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
Check tokens...
Selected tokens are correct
  0% 1/2000 [03:07<77:33:22, 139.67s/it, loss=0.0629]`



it takes : 21.5 / 40.0 GB VRAM for textual inversion. so we ca meet during the next following weekend and run tests on my card I have enough VRAM to run it

**Reply by seruva19** [reply]
Understood, thanks for checking.

**Comment by adrianpuiu** [comment]
https://drive.google.com/file/d/1-gJOyxqL1rJBnns9GpzoytJCwouUrfgw/view?usp=sharing , try this oil painting embedding , keyword  : sks , trained on 10 images 100 epochs => almost 4 hours

**Reply by seruva19** [reply]
Thank you for your efforts!
~~On free colab I get CUDA OOM error when running txt2img on a model with loaded embedding.~~

Sorry for misinformation, inference actually works with 16 Gb Tesla T4, total VRAM usage does not exceed ~9.5 Gb.
Got so used to batch_size=1 that didn't notice I ran notebook with batch_size=4.
I will test the embedding and post images here later, just a little bit busy rn.

**Reply by adrianpuiu** [reply]
They should come out as oil painting style with heavy strokes , paint dripping, something similar with the ones on playground.ai

**Reply by seruva19** [reply]
For inference I used slightly adapted [notebook](https://github.com/TheDenk/Kandinsky-2-textual-inversion/blob/main/notebooks/inference_textual_inversion.ipynb) from https://github.com/TheDenk/Kandinsky-2-textual-inversion.
The prompt was: `Professional high-quality art of a sks. photorealistic, 4k, HQ` with num_steps=50, batch_size=1, guidance_scale=7.5, h=768, w=768, sampler='p_sampler', prior_cf_scale=2, prior_steps='4'.

<details> 
<summary>Output for model without embeddings</summary>

![0](https://user-images.githubusercontent.com/26826215/234302780-7a158e28-c3f8-484b-826e-53344b4d6358.png)

![1](https://user-images.githubusercontent.com/26826215/234302863-3291fc3e-f949-47d3-a26f-173034d74b65.png)

![2](https://user-images.githubusercontent.com/26826215/234302891-0eca8e0b-68c4-415e-b4cd-677f3706b9e5.png)

![3](https://user-images.githubusercontent.com/26826215/234302911-d80a63e9-3e4e-44ad-a09e-3ce4a97fc4e1.png)
</details> 

<details> 
<summary>Output for model with your embeddings</summary>

![0 (1)](https://user-images.githubusercontent.com/26826215/234302956-8025a078-e879-46ec-b571-475518f481d9.png)

![1 (1)](https://user-images.githubusercontent.com/26826215/234303007-6e558906-c9e9-4e19-a1af-235ad6edf176.png)

![2 (1)](https://user-images.githubusercontent.com/26826215/234303053-84f2a83b-f3b4-43f7-98c6-487612721bea.png)

![3 (1)](https://user-images.githubusercontent.com/26826215/234303068-c7e60f4a-c82a-4791-b784-0f874ea05f30.png)
</details> 

Looks like it works ;)

**Comment by adrianpuiu** [comment]
It works but it’s not that I trained it with .. or the training went wrong
… did you use the textual inversion from the Google drive ?



On Tue 25. 4. 2023 at 16:17, neutron_hare ***@***.***> wrote:

> For inference I used slightly adapted notebook
> <https://github.com/TheDenk/Kandinsky-2-textual-inversion/blob/main/notebooks/inference_textual_inversion.ipynb>
> from https://github.com/TheDenk/Kandinsky-2-textual-inversion.
> The prompt was: Professional high-quality art of a sks. photorealistic,
> 4k, HQ with num_steps=50, batch_size=1, guidance_scale=7.5, h=768, w=768,
> sampler='p_sampler', prior_cf_scale=2, prior_steps='4'.
> Output for model without embeddings
>
> [image: 0]
> <https://user-images.githubusercontent.com/26826215/234302780-7a158e28-c3f8-484b-826e-53344b4d6358.png>
>
> [image: 1]
> <https://user-images.githubusercontent.com/26826215/234302863-3291fc3e-f949-47d3-a26f-173034d74b65.png>
>
> [image: 2]
> <https://user-images.githubusercontent.com/26826215/234302891-0eca8e0b-68c4-415e-b4cd-677f3706b9e5.png>
>
> [image: 3]
> <https://user-images.githubusercontent.com/26826215/234302911-d80a63e9-3e4e-44ad-a09e-3ce4a97fc4e1.png>
> Output for model with your embeddings
>
> [image: 0 (1)]
> <https://user-images.githubusercontent.com/26826215/234302956-8025a078-e879-46ec-b571-475518f481d9.png>
>
> [image: 1 (1)]
> <https://user-images.githubusercontent.com/26826215/234303007-6e558906-c9e9-4e19-a1af-235ad6edf176.png>
>
> [image: 2 (1)]
> <https://user-images.githubusercontent.com/26826215/234303053-84f2a83b-f3b4-43f7-98c6-487612721bea.png>
>
> [image: 3 (1)]
> <https://user-images.githubusercontent.com/26826215/234303068-c7e60f4a-c82a-4791-b784-0f874ea05f30.png>
>
> Looks like it works ;)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/seruva19/kubin/discussions/10#discussioncomment-5720649>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABNVRVAGSQOHIFKRDJKPJ3XC7MGVANCNFSM6AAAAAAXJE66X4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Best regards,
George

**Reply by seruva19** [reply]
Yes, exactly. I used this link for direct download https://drive.google.com/uc?id=1-gJOyxqL1rJBnns9GpzoytJCwouUrfgw
What should the result look like? Can you give few examples of images you used for training?

**Comment by adrianpuiu** [comment]
These are the images I trained it with .. here’s the link :
https://drive.google.com/file/d/1-0lToiMrU7xS7Ib6sIRJi7suT08sKr-9/view?usp=drivesdk



On Tue 25. 4. 2023 at 17:49, neutron_hare ***@***.***> wrote:

> Yes, exactly. I used this link for direct download
> https://drive.google.com/uc?id=1-gJOyxqL1rJBnns9GpzoytJCwouUrfgw
> What should the result look like? Can you give few examples of images you
> used for training?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/seruva19/kubin/discussions/10#discussioncomment-5721794>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABNVRRR7DP4AS26YCCAE4DXC7XA5ANCNFSM6AAAAAAXJE66X4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Best regards,
George

**Comment by adrianpuiu** [comment]
Wati , i sunt YOU The wrong link… try this one


https://drive.google.com/drive/folders/1CCUFrmwjmZQv30GDsHgOenPjzKMSzwZi

On Tue 25. 4. 2023 at 17:49, neutron_hare ***@***.***> wrote:

> Yes, exactly. I used this link for direct download
> https://drive.google.com/uc?id=1-gJOyxqL1rJBnns9GpzoytJCwouUrfgw
> What should the result look like? Can you give few examples of images you
> used for training?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/seruva19/kubin/discussions/10#discussioncomment-5721794>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABNVRRR7DP4AS26YCCAE4DXC7XA5ANCNFSM6AAAAAAXJE66X4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Best regards,
George

**Reply by seruva19** [reply]
What _initializer_token_ did you use for training?

**Comment by adrianpuiu** [comment]
If I remember correctly I think it was : painting . Try with this one… I’ll
open the collar in a min and check
On Tue 25. 4. 2023 at 21:35, neutron_hare ***@***.***> wrote:

> What *initializer_token* did you use for training?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/seruva19/kubin/discussions/10#discussioncomment-5723945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABNVRUWZ5A27XKUPCU33Q3XDARQRANCNFSM6AAAAAAXJE66X4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Best regards,
George

**Reply by seruva19** [reply]
Tried another prompt: `sks painting` The results are weird:

<details>
<summary>Without embeddings</summary>

![0](https://user-images.githubusercontent.com/26826215/234393214-570b4ec1-b78b-40ce-9dbd-2488e82fc6ce.png)

![1](https://user-images.githubusercontent.com/26826215/234393265-2a3be88d-bfbc-45b3-aefd-048af82ea2e7.png)

</details>

<details>
<summary>With embeddings</summary>

![0 (1)](https://user-images.githubusercontent.com/26826215/234393471-06ceecc0-ce47-49ef-af03-328e88819c76.png)

![1 (1)](https://user-images.githubusercontent.com/26826215/234393540-66e759f2-c4a3-416b-8476-879a88a7275f.png)

<details>

**Comment by adrianpuiu** [comment]
There’s a guy modifying it to diffusers and now that nvidia published
tensorrt code to  huggingface diffusers , if he sees it , he might get it
speeded up about 4 5 times then usual …

On Tue 25. 4. 2023 at 21:37, Adrian Puiu ***@***.***> wrote:

>
> If I remember correctly I think it was : painting . Try with this one…
> I’ll open the collar in a min and check
>
> On Tue 25. 4. 2023 at 21:35, neutron_hare ***@***.***>
> wrote:
>
>> What *initializer_token* did you use for training?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/seruva19/kubin/discussions/10#discussioncomment-5723945>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AABNVRUWZ5A27XKUPCU33Q3XDARQRANCNFSM6AAAAAAXJE66X4>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
> --
> Best regards,
> George
>
-- 
Best regards,
George

**Reply by seruva19** [reply]
You mean https://github.com/huggingface/diffusers/issues/2985#issuecomment-1521518672?
I also hope they would succeed. I started writing diffusers wrapper for Kandinsky myself, but my skills are not high enough for this :(

Speedup would be nice, but it won't solve memory problem. There are two diffusion models in Kandinsky pipeline (prior and unet), together they do not fit into 8 Gb. And I was hoping to somehow find a way to optimize and make it "consumer-friendly", so it might be able to compete with Stable Diffusion. But for this it has to be accessible for users without high-end GPUs.

**Comment by adrianpuiu** [comment]
:)) those look weird :)) ..  I made a few last night but got them on pc,
they came out really nice, don’t forget the text decoder works better then
in stable diffusion. Give it something like this: a beautiful painting of a
 meadow covered with grass and poppies and tall pine lit by sun at golden
hour.






On Tue 25. 4. 2023 at 22:20, neutron_hare ***@***.***> wrote:

> Tried another prompt: sks painting The results are weird:
> Without embeddings
>
> [image: 0]
> <https://user-images.githubusercontent.com/26826215/234393214-570b4ec1-b78b-40ce-9dbd-2488e82fc6ce.png>
>
> [image: 1]
> <https://user-images.githubusercontent.com/26826215/234393265-2a3be88d-bfbc-45b3-aefd-048af82ea2e7.png>
> With embeddings
>
> [image: 0 (1)]
> <https://user-images.githubusercontent.com/26826215/234393471-06ceecc0-ce47-49ef-af03-328e88819c76.png>
>
> [image: 1 (1)]
> <https://user-images.githubusercontent.com/26826215/234393540-66e759f2-c4a3-416b-8476-879a88a7275f.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/seruva19/kubin/discussions/10#discussioncomment-5724301>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABNVRUUZ2SPP4BEJXWAXLLXDAWZLANCNFSM6AAAAAAXJE66X4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Best regards,
George

**Reply by seruva19** [reply]
Thank you, I will try tomorrow.

**Comment by seruva19** [comment]
Still not getting it right...
`a beautiful painting of a sks in a forest lit with noonday sun with multiple animals walking around`

<details><summary>Embedding OFF</summary>

![3](https://user-images.githubusercontent.com/26826215/234592655-4052f4da-f839-40b0-9a93-55cef5f1a3a6.png)
</details>

<details><summary>Embedding ON</summary>

![3 (1)](https://user-images.githubusercontent.com/26826215/234592690-41ebb0ac-7e23-421c-9c6c-a994c722e317.png)
</details>

Tbf, I never trained a textual inversion so I have no idea what's wrong with your embeddings (if there is an error).

I will add UI for training into my app, but right now I am working on extensions support, upscaler and image browser components, after that gonna move on to training widgets and hopefully do some additional research afterwards.

Btw, I am very interested in approaches suggested in the paper referenced [here](https://www.reddit.com/r/StableDiffusion/comments/12yzd2a/google_researchers_achieve_performance/), looks promising.

**Reply by adrianpuiu** [reply]
class CoreAttention(MegatronModule):  https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/transformer.py#L203-L209

**Reply by seruva19** [reply]
Thanks, will take a look.

**Comment by seruva19** [comment]
DeepFloyd-IF was [just released](https://github.com/deep-floyd/IF) and there is a [blog post on 🤗](https://huggingface.co/blog/if) that contains some insights into optimizing model for low-end GPUs and its usage with diffusers (and probably some of them can be used for Kandinsky as well). Guess I found some reading for next weeks :)

**Comment by adrianpuiu** [comment]
hi buddy, hope you're ok. 
Could you also add on your list :  custom gr box for creating / removing own created styles in the t2i tab/ something similar with the ones a1111 has

**Reply by seruva19** [reply]
Hi. Well, actually there is already a built-in extension that allows usage of predefined styles, and I think I can tweak it to enable editing user styles:

![image](https://user-images.githubusercontent.com/26826215/235749872-7908ff2a-9163-429d-9ce4-aab1519d9ad4.png)

Perhaps I might do it near weekend.

**Comment by adrianpuiu** [comment]
perfect. if  i get some time this ill try to modify the t2i ui tab and
create more of a image generation workflow similar to the one i had on my
little t2i app  and try to get more of that gallery space on screen. it
only needs an accordion that get hidden once the image settings where set
cause the majority of them generate at fixed size then upscale so no point
in keeping  them to occupy that space that all about .. generating and
viewing the images

**Reply by seruva19** [reply]
Yeah, I know, the gallery panel and UI in general needs to be improved, and at this point I don't take care of UX enough 😀

**Comment by adrianpuiu** [comment]
i'll try to do it. you focus on the coding part. :)) i've tried to
understand the code but neah, its not mine if you know what i mean :)). aw
btw make a template for extensions  when you have time , maybe others will
join and help

**Reply by seruva19** [reply]
I will generate extension template for sure, just not finalized its generic implementation and API, so not ready yet to expose it. Once I get done outpainting and training widgets (the latter will also be a built-in extension), I'll add dev templates, unit tests and probably some sort of docs.

**Comment by adrianpuiu** [comment]
there's a weird bug in Gradio when using gr.Box() 

`OSError: [Errno 98] Address already in use

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/agp/kubin/src/kubin.py", line 39, in <module>
    ui.queue(concurrency_count=kubin.args.concurrency_count, api_open=False).launch(
  File "/home/agp/kubin/venv/lib/python3.10/site-packages/gradio/blocks.py", line 1512, in launch
    server_name, server_port, local_url, app, server = networking.start_server(
  File "/home/agp/kubin/venv/lib/python3.10/site-packages/gradio/networking.py", line 122, in start_server
    raise OSError(
OSError: Port 7860 is in use. If a gradio.Blocks is running on the port, you can close() it or gradio.close_all().`

**Comment by kodxana** [comment]
Official training notebooks came out :)
https://github.com/ai-forever/Kandinsky-2/tree/main/notebooks

**Reply by seruva19** [reply]
Thanks! I will use use these notebooks and also scripts from https://github.com/TheDenk/Kandinsky-2-textual-inversion

**Reply by kodxana** [reply]
Some notes:
On A6000 48 GB GPU I was able to train prio model with batch size 140
and for unclip model batch size 4
Here is sample of my finetune model:
![image](https://github.com/seruva19/kubin/assets/16674412/5f009326-ed94-41a4-8384-6244dd194f5f)

Also here is script for preparing dataset for training:

```python
import os
import pandas as pd
from PIL import Image

path_to_images_directory = '/workspace/data'

# Collect image filenames and captions
data = []
for filename in os.listdir(path_to_images_directory):
    if filename.endswith(('.jpg', '.jpeg', '.png', '.bmp')):
        # Assume the caption is the part of the filename before the extension
        caption = os.path.splitext(filename)[0]
        image_path = os.path.join(path_to_images_directory, filename)
        data.append([image_path, caption])

# Create a DataFrame and save it as a CSV
df = pd.DataFrame(data, columns=['image_name', 'caption'])
df.to_csv('test.csv', index=False)

# Create a new directory named 'processed_images'
os.mkdir('images')

# Process images and save them in the 'processed_images' directory
for image_path, caption in data:
    image = Image.open(image_path)
    resized_image = image.resize((768, 768))  # Resize the image to 768x768 pixels
    new_image_path = os.path.join('images', os.path.basename(image_path))
    resized_image.save(new_image_path)

**Reply by seruva19** [reply]
Btw this must be your post here: https://www.reddit.com/r/CodeLyoko/comments/13geunw/japanese_aelita_ai_art/ ?
Do you use runpod for training?

**Reply by kodxana** [reply]
yes and yes

**Comment by seruva19** [comment]
Finished (for now) with outpainting and smaller stuff and will start working on extension for fine-tuning :)
I am pretty impressed and motivated with the results of your work @adrianpuiu (I mean the [recent reddit post](https://www.reddit.com/r/StableDiffusion/comments/13hgpo2/kandinsky_21_fine_tune/)).
I believe the base Kandinsky model is good enough to be the foundation for multiple MJ-quality community models. But first we need GUI for fine-tuning + some kind of ControlNet + some optimizations concerning VRAM consumption (diffusers implementation will give these opportunities).

**Reply by adrianpuiu** [reply]
thank you. i really like this model more then the autistic stable diffusion and the way it understands text. for the next few days ill just stay on the midjourney showcase feed and build my dataset then find a way to make that docker image work and start the training , if it wont work ill just make the training on my pc , i just need to find out how long it took the other guy to train both models for those 650 mentioned photos and make a rough timing assumption  . 

PS : can't wait for those training tabs :D . IF they finish with diffuers you might have a good chance to controlnet, vram optimisation will come by default

**Reply by kodxana** [reply]
> thank you. i really like this model more then the autistic stable diffusion and the way it understands text. for the next few days ill just stay on the midjourney showcase feed and build my dataset then find a way to make that docker image work and start the training , if it wont work ill just make the training on my pc , i just need to find out how long it took the other guy to train both models for those 650 mentioned photos and make a rough timing assumption .
> 
> PS : can't wait for those training tabs :D . IF they finish with diffuers you might have a good chance to controlnet, vram optimisation will come by default

For training I used A6000 48 GB VRAM. Trained on RunPod :)

**Reply by adrianpuiu** [reply]
ill do the training on my pc. hopefully nobody will unplug it :))

**Comment by seruva19** [comment]
I have published an extension for training prior and unclip models on the 'dev' branch. It includes a GUI for preparing datasets and training both models.

Technically, it works fine as I have tested it on free Colab and Paperspace (where I have a pro subscription). By _technically_, I mean that I was able to successfully train models and obtain checkpoints. However, I have not tested (yet) the image output from these checkpoints 😉

Before merging it into the main branch, I need to perform more testing and add additional necessary functions. Currently, there is no UI for applying a trained checkpoint, and the original Kandinsky checkpoints need to be replaced manually. Additionally, the UI feels somewhat bloated, and I need to address that. I would greatly appreciate feedback from @adrianpuiu and @kodxana who have already trained Kandinsky via notebooks. Any ideas or suggestions to make the UI more convenient would be highly valued :)

<details><summary>Screenshots</summary>


![image](https://github.com/seruva19/kubin/assets/26826215/ba650e5c-7579-40e2-9a5f-d29ea6a3e4b3)
![image](https://github.com/seruva19/kubin/assets/26826215/4b917786-35e0-4253-a8f4-cd0cdfce3ae2)
![image](https://github.com/seruva19/kubin/assets/26826215/d8434ab5-2d20-4dd9-b5de-6efc0b6a529b)

</details>

**Reply by kodxana** [reply]
So far looks good will give it a spin, capition editor would be great thing.

**Reply by seruva19** [reply]
Yes, I plan to add editing captions via GUI as well.

**Reply by adrianpuiu** [reply]
tested it so far and its great I don't see anything missing . apart from what @kodxana  said , a captioning editor but we can do without it for a while. better focus on some new  feature or enhancement/plugin .

![Uploading Screenshot from 2023-05-28 22-58-03.png…]()

**Comment by adrianpuiu** [comment]
@seruva19  when you have some time. can you please add a clip interrogator for creating captions from data set folder ?

**Reply by seruva19** [reply]
Yes, I was thinking about it already.

**Reply by adrianpuiu** [reply]
i finally found a way to scrape the mid journey for pictures and that was the only thing missing ... now i have to think of how the heck im going to be using jquery class selector to extract the captions from their website ... a tool like that would be awesome to have even before preparing the dataset in kubin .

**Reply by seruva19** [reply]
Are you trying to scrape their community showcase https://www.midjourney.com/showcase/ ?

**Reply by adrianpuiu** [reply]
yep. they have around 320+ images / day .. im extracting them using https://extract.pics but the only downside is that they allow you to select and download around 100 images at once .. and  not having the image caption is their only downfall ..i ased few days ago some bulgarian guys if they would share their 700K+ image data set, they used same method using 2 corporate midjourney accounts

**Reply by Blucknote** [reply]
> i finally found a way to scrape the mid journey for pictures and that was the only thing missing ... now i have to think of how the heck im going to be using jquery class selector to extract the captions from their website ... a tool like that would be awesome to have even before preparing the dataset in kubin .

try this
```javascript
document.querySelectorAll('img').forEach(function(img) {
  console.log(img.alt);
  console.log(img.src);
  console.log("=============")
});
```

**Reply by adrianpuiu** [reply]
you're a butt saver man. thanks

**Reply by seruva19** [reply]
Slightly modified script made by @Blucknote so it would filter out non-relevant images and output the result in single array:

```
console.log(Array.from(document.querySelectorAll('img')).map(img => (
  {img: img.src, caption: img.alt} 
)).filter(({caption}) => !!caption))
```

**Reply by seruva19** [reply]
@adrianpuiu I added GUI for batch processing in interrogation tab.

**Reply by adrianpuiu** [reply]
aw man, sweet. thank you!

**Reply by adrianpuiu** [reply]
330it/s * 86 pics  done , splendid work. loving it

**Comment by adrianpuiu** [comment]
so im trying to work around something like this : http://jsfiddle.net/Snfst/6/

**Reply by adrianpuiu** [reply]
https://aituts.com/find-midjourney-user/?fbclid=IwAR2PWMK2AuC4Y4j5U22wmBcnfdwJH3DU3F6vNIxwwfFNOIckljct4JwKUL0
